Dockerfile

```diff
@@ -1,294 +1,550 @@
-# an older (pre v2.3.0) version of the registry that only supports schema1
-# manifests. This version of the registry is not working on arm64, so installation
-# is skipped on that architecture.
-ARG REGISTRY_VERSION_SCHEMA1=v2.1.0
+RUN --mount=type=cache,sharing=locked,id=moby-base-aptlib,target=/var/lib/apt \
+    --mount=type=cache,sharing=locked,id=moby-base-aptcache,target=/var/cache/apt \
+    apt-get update && apt-get install --no-install-recommends -y \
+      bash \
+      ca-certificates \
+      cmake \
+      file \
+      gcc \
+      git \
+      libc6-dev \
+      lld \
+      make \
+      pkg-config
+COPY --from=golang /usr/local/go /usr/local/go
+ENV GOROOT="/usr/local/go"
+ENV GOPATH="/go"
+ENV PATH="$GOPATH/bin:/usr/local/go/bin:$PATH"
+RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
+
+# go-winres
+FROM base AS gowinres
+ARG GOWINRES_VERSION
 RUN --mount=type=cache,target=/root/.cache/go-build \
     --mount=type=cache,target=/go/pkg/mod \
-    --mount=type=tmpfs,target=/go/src/ \
-        set -x \
-        && git clone https://github.com/docker/distribution.git . \
-        && git checkout -q "$REGISTRY_VERSION" \
-        && GOPATH="/go/src/github.com/docker/distribution/Godeps/_workspace:$GOPATH" \
-           go build -buildmode=pie -o /build/registry-v2 github.com/docker/distribution/cmd/registry \
-        && case $(dpkg --print-architecture) in \
-               amd64|armhf|ppc64*|s390x) \
-               git checkout -q "$REGISTRY_VERSION_SCHEMA1"; \
-               GOPATH="/go/src/github.com/docker/distribution/Godeps/_workspace:$GOPATH"; \
-                   go build -buildmode=pie -o /build/registry-v2-schema1 github.com/docker/distribution/cmd/registry; \
-                ;; \
-           esac
+    GOBIN=/out GO111MODULE=on go install "github.com/tc-hib/go-winres@${GOWINRES_VERSION}" \
+    && /out/go-winres --help
 
-FROM base AS swagger
-WORKDIR $GOPATH/src/github.com/go-swagger/go-swagger
-
-# GO_SWAGGER_COMMIT specifies the version of the go-swagger binary to build and
-# install. Go-swagger is used in CI for validating swagger.yaml in hack/validate/swagger-gen
-#
-# Currently uses a fork from https://github.com/kolyshkin/go-swagger/tree/golang-1.13-fix,
-# TODO: move to under moby/ or fix upstream go-swagger to work for us.
-ENV GO_SWAGGER_COMMIT c56166c036004ba7a3a321e5951ba472b9ae298c
-RUN --mount=type=cache,target=/root/.cache/go-build \
-    --mount=type=cache,target=/go/pkg/mod \
-    --mount=type=tmpfs,target=/go/src/ \
-        set -x \
-        && git clone https://github.com/kolyshkin/go-swagger.git . \
-        && git checkout -q "$GO_SWAGGER_COMMIT" \
-        && go build -o /build/swagger github.com/go-swagger/go-swagger/cmd/swagger
+# containerd
+FROM base AS containerd-src
+WORKDIR /usr/src/containerd
+RUN git clone https://github.com/containerd/containerd.git .
+ARG CONTAINERD_VERSION
+RUN git fetch origin && git checkout -q "$CONTAINERD_VERSION"
```

Comment:
Yeah, perhaps. Although if we copy it to one of the base images, it would only be invalidated if that script itself changed, correct?

I was curious; did a quick try; looks to be faster (but I guess, it depends also on how many of these we're running) (would be nice if the output showed size as well, besides timing)

```dockerfile
# syntax=docker/dockerfile:1

ARG CONTAINERD_VERSION=v1.6.6

FROM golang:1.18-bullseye AS golang

FROM golang AS clone
WORKDIR /usr/src/containerd
ARG CONTAINERD_VERSION
RUN git clone https://github.com/containerd/containerd.git . \
 && git fetch origin && git checkout -q "$CONTAINERD_VERSION"
 
FROM golang AS shallow1
WORKDIR /usr/src/containerd
ARG CONTAINERD_VERSION
RUN <<EOF
set -ex

git init .
git remote add origin "https://github.com/containerd/containerd.git"

REF="$CONTAINERD_VERSION"
REF_FETCH="$REF"
# if ref is branch or tag, retrieve its canonical form
REF=$(git ls-remote --refs --heads --tags origin "$REF" | awk '{print $2}')
if [ -n "$REF" ]; then
    # if branch or tag then create it locally too
    REF_FETCH="$REF:$REF"
else
    REF="FETCH_HEAD"
fi
git fetch --update-head-ok --depth 1 origin "$REF_FETCH"
git checkout -q "$REF"
EOF


FROM golang AS shallow2
WORKDIR /usr/src/containerd
ARG CONTAINERD_VERSION
RUN <<EOF
git init .
git remote add origin "https://github.com/containerd/containerd.git"
git fetch --depth 1 origin "$(CONTAINERD_VERSION)"
git checkout -q FETCH_HEAD
EOF

FROM scratch AS final
COPY --from=clone    /usr/src/containerd/Makefile .
COPY --from=shallow1 /usr/src/containerd/Makefile .
COPY --from=shallow2 /usr/src/containerd/Makefile .
```

```bash
docker build -t foo --no-cache .
[+] Building 14.1s (16/16) FINISHED
 => [internal] load build definition from Dockerfile                                                                                    0.0s
 => => transferring dockerfile: 32B                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                       0.0s
 => => transferring context: 2B                                                                                                         0.0s
 => resolve image config for docker.io/docker/dockerfile:1                                                                              0.9s
 => CACHED docker-image://docker.io/docker/dockerfile:1@sha256:443aab4ca21183e069e7d8b2dc68006594f40bddf1b15bbd83f5137bd93e80e2         0.0s
 => [internal] load build definition from Dockerfile                                                                                    0.0s
 => [internal] load .dockerignore                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/golang:1.18-bullseye                                                                 0.9s
 => [golang 1/1] FROM docker.io/library/golang:1.18-bullseye@sha256:6bc0fe859a2a28af025f8885fc073996dd14d0371d5c0474aed798317849bc18    0.0s
 => => resolve docker.io/library/golang:1.18-bullseye@sha256:6bc0fe859a2a28af025f8885fc073996dd14d0371d5c0474aed798317849bc18           0.0s
 => CACHED [shallow2 1/2] WORKDIR /usr/src/containerd                                                                                   0.0s
 => [clone 2/2] RUN git clone https://github.com/containerd/containerd.git .  && git fetch origin && git checkout -q "v1.6.6"          11.8s
 => [shallow2 2/2] RUN <<EOF (git init ....)                                                                                            6.0s
 => [shallow1 2/2] RUN <<EOF (set -ex...)                                                                                               7.7s
 => [final 1/3] COPY --from=clone    /usr/src/containerd/Makefile .                                                                     0.0s
 => [final 2/3] COPY --from=shallow1 /usr/src/containerd/Makefile .                                                                     0.0s
 => [final 3/3] COPY --from=shallow2 /usr/src/containerd/Makefile .                                                                     0.0s
 => exporting to image                                                                                                                  0.0s
 => => exporting layers                                                                                                                 0.0s
 => => writing image sha256:a22d3ddcbdbc94d5f5f1f484ee40527474be7d4ea46d8a87146955ea7dcf70ff                                            0.0s
 => => naming to docker.io/library/foo                                                                                                  0.0s

```

Reply:
Yes indeed with `--depth=1` it's faster